src/typeChecker/typeErrorChecker.ts

```diff
@@ -1109,34 +1249,86 @@ function getAnnotatedType(typeNode: tsEs.TSType, context: Context): Type {
       }
       return tLiteral(value)
     case 'TSArrayType':
-      return tArray(getAnnotatedType(typeNode.elementType, context))
+      return tArray(getAnnotatedType(typeNode.elementType))
     case 'TSUnionType':
-      const unionTypes = typeNode.types.map(node => getAnnotatedType(node, context))
-      return mergeTypes(...unionTypes)
+      const unionTypes = typeNode.types.map(node => getAnnotatedType(node))
+      return mergeTypes(typeNode, ...unionTypes)
     case 'TSIntersectionType':
       throw new TypecheckError(typeNode, 'Intersection types are not allowed')
     case 'TSTypeReference':
       const name = typeNode.typeName.name
-      return lookupTypeAliasAndRemoveForAllAndPredicateTypes(typeNode, name, context)
+      // Save name and type arguments in variable type
+      if (typeNode.typeParameters) {
+        const typesToSub: Type[] = []
+        for (const paramNode of typeNode.typeParameters.params) {
+          if (paramNode.type === 'TSTypeParameter') {
+            throw new TypecheckError(typeNode, 'Type argument should not be type parameter')
+          }
+          typesToSub.push(getAnnotatedType(paramNode))
+        }
+        return tVar(name, typesToSub)
+      }
+      return tVar(name)
     case 'TSParenthesizedType':
-      return getAnnotatedType(typeNode.typeAnnotation, context)
+      return getAnnotatedType(typeNode.typeAnnotation)
     default:
-      return getBasicType(typeNode, context)
+      return getBasicType(typeNode)
   }
 }
 
 /**
- * Converts array of function parameters into array of types.
+ * Converts an array of function parameters into an array of types.
+ */
+function getParamTypes(params: (tsEs.Identifier | tsEs.RestElement)[]): Type[] {
+  return params.map(param => getTypeAnnotationType(param.typeAnnotation))
+}
+
+/**
+ * Returns the head type of the input type.
  */
-function getParamTypes(params: (tsEs.Identifier | tsEs.RestElement)[], context: Context): Type[] {
-  return params.map(param => getTypeAnnotationType(param.typeAnnotation, context))
+function getHeadType(node: tsEs.Node, type: Type): Type {
+  switch (type.kind) {
+    case 'pair':
+      return type.headType
+    case 'list':
+      return type.elementType
+    case 'union':
+      return tUnion(...type.types.map(type => getHeadType(node, type)))
```

Comment:
`type.types` seems weird. `type.constituentTypes`?

Reply:
On second thoughts, will be keeping the attribute name as `types` to be consistent with the naming in the `TSUnionType` node.